UnitTests/UnitTests.HighPerformance.Shared/Memory/Test_Memory2D{T}.cs

```diff
@@ -0,0 +1,529 @@
+        {
+            int[,] array =
+            {
+                { 1, 2, 3 },
+                { 4, 5, 6 }
+            };
+
+            // Test the constructor taking a T[,] array that is mapped directly (no slicing)
+            Memory2D<int> memory2d = new Memory2D<int>(array);
+
+            Assert.IsFalse(memory2d.IsEmpty);
+            Assert.AreEqual(memory2d.Length, 6);
+            Assert.AreEqual(memory2d.Width, 3);
+            Assert.AreEqual(memory2d.Height, 2);
+            Assert.AreEqual(memory2d.Span[0, 1], 2);
+            Assert.AreEqual(memory2d.Span[1, 2], 6);
+
+            // Here we test the check for covariance: we can't create a Memory2D<T> from a U[,] array
+            // where U is assignable to T (as in, U : T). This would cause a type safety violation on write.
+            Assert.ThrowsException<ArrayTypeMismatchException>(() => new Memory2D<object>(new string[1, 2]));
+        }
+
+        [TestCategory("Memory2DT")]
+        [TestMethod]
+        public void Test_Memory2DT_Array2DConstructor_2()
+        {
+            int[,] array =
+            {
+                { 1, 2, 3 },
+                { 4, 5, 6 }
+            };
+
+            // Same as above, but this time we also slice the memory to test the other constructor
+            Memory2D<int> memory2d = new Memory2D<int>(array, 0, 1, 2, 2);
+
+            Assert.IsFalse(memory2d.IsEmpty);
+            Assert.AreEqual(memory2d.Length, 4);
+            Assert.AreEqual(memory2d.Width, 2);
+            Assert.AreEqual(memory2d.Height, 2);
+            Assert.AreEqual(memory2d.Span[0, 0], 2);
+            Assert.AreEqual(memory2d.Span[1, 1], 6);
+
+            Assert.ThrowsException<ArrayTypeMismatchException>(() => new Memory2D<object>(new string[1, 2], 0, 0, 2, 2));
+        }
+
+        [TestCategory("Memory2DT")]
+        [TestMethod]
+        public void Test_Memory2DT_Array3DConstructor_1()
+        {
+            int[,,] array =
+            {
+                {
+                    { 1, 2, 3 },
+                    { 4, 5, 6 }
+                },
+                {
+                    { 10, 20, 30 },
+                    { 40, 50, 60 }
+                }
+            };
+
+            // Same as above, but we test the constructor taking a layer within a 3D array
+            Memory2D<int> memory2d = new Memory2D<int>(array, 1);
+
+            Assert.IsFalse(memory2d.IsEmpty);
+            Assert.AreEqual(memory2d.Length, 6);
+            Assert.AreEqual(memory2d.Width, 3);
+            Assert.AreEqual(memory2d.Height, 2);
+            Assert.AreEqual(memory2d.Span[0, 1], 20);
+            Assert.AreEqual(memory2d.Span[1, 2], 60);
+
+            // A couple of tests for invalid parameters, ie. layers out of range
+            Assert.ThrowsException<ArgumentOutOfRangeException>(() => new Memory2D<int>(array, -1));
+            Assert.ThrowsException<ArgumentOutOfRangeException>(() => new Memory2D<int>(array, 20));
```

Comment:
same here for off-by-one

Reply:
Added a few new tests for this and `ReadOnlyMemory2D<T>` versions in 7b04c7d4a7846ddaff3f6686307984c8970fb696 👍